File: Tyche.Android/ui/src/main/java/com/felipearpa/ui/paging/CursorPagingSource.kt
package com.felipearpa.ui.paging

import androidx.paging.PagingSource
import androidx.paging.PagingState
import com.felipearpa.core.paging.CursorPage

typealias CursorPagingQuery<T> = suspend (String?) -> Result<CursorPage<T>>

open class CursorPagingSource<T : Any>(private val pagingQuery: CursorPagingQuery<T>) :
    PagingSource<String, T>() {

    override suspend fun load(params: LoadParams<String>): LoadResult<String, T> {
        return pagingQuery(params.key).fold(
            onSuccess = { page ->
                LoadResult.Page(
                    data = page.items,
                    prevKey = null,
                    nextKey = page.next
                )
            },
            onFailure = { exception ->
                LoadResult.Error(exception)
            }
        )
    }

    override fun getRefreshKey(state: PagingState<String, T>): String? = null
}

File: Tyche.iOS/UI/Sources/UI/Paging/CursorPagingSource.swift
import Core

public typealias CursorPagingQuery<Item: Codable> = (String?) async -> Result<CursorPage<Item>, Error>

open class CursorPagingSource<TItem: Codable>: PagingSource<String, TItem> {
    public typealias TKey = String
    public typealias TItem = TItem
    
    private let pagingQuery: CursorPagingQuery<TItem>
    
    public init(pagingQuery: @escaping CursorPagingQuery<TItem>) {
        self.pagingQuery = pagingQuery
    }
    
    public override func load(loadConfig: LoadConfig<TKey>) async -> LoadResult<TKey, TItem> {
        let result = await pagingQuery(loadConfig.key)
        return result.fold(
            onSuccess: { page in
                LoadResult.page(items: page.items, nextKey: page.next)
            },
            onFailure: { error in
                LoadResult.failure(error)
            }
        )
    }
}
